Add GPU autoscheduler

aekul opened the pull request:
This is a draft PR to merge the anderson2021 GPU autoscheduler (#5602). It includes the code (there is some overlap with adams2019, but it will probably be a substantial amount of work to deduplicate them), some tests, utility scripts for generating data/statistics/etc. (these can be removed if desired), and baseline weights (trained on a V100).

steven-johnson commented on src/autoschedulers/anderson2021/Makefile:
```diff
@@ -0,0 +1,292 @@
+$(BIN)/%/demo.a: $(GENERATOR_BIN)/demo.generator $(BIN)/libautoschedule_anderson2021.$(SHARED_EXT)
+	@mkdir -p $(@D)
+	HL_WEIGHTS_DIR=$(SRC)/baseline.weights \
+	$(GENERATOR_BIN)/demo.generator -g demo -o $(@D) -f demo target=$* auto_schedule=true -p $(BIN)/libautoschedule_anderson2021.$(SHARED_EXT) -s Anderson2021
```
Here and elsewhere: `auto_schedule=true -s Anderson2021` is no longer supported syntax. Instead, specify `autoscheduler=Anderson2021` to replace both of these.

steven-johnson reviewed (commented):
Yeah, there really is a *lot* of overlap with Adams2019 here. Would it be tractable to merge these changes into Adams2019, or at least have shared code between them?

aekul commented:
Some of the files (e.g. `ASLog.h/.cpp`, `PerfectHashMap.h`, and a few others) are identical and could be merged. Would `autoschedulers/common` be the right place for them? Looks like `ASLog.h/.cpp` are already there.

Unfortunately, most of the other files (e.g. `Autoschedule.h/cpp`, `LoopNest.h/cpp`, etc.) have diverged considerably and it would be a significant amount of work to merge them.

steven-johnson commented:
> Some of the files (e.g. `ASLog.h/.cpp`, `PerfectHashMap.h`, and a few others) are identical and could be merged. Would `autoschedulers/common` be the right place for them? Looks like `ASLog.h/.cpp` are already there.

Yep!
 
> Unfortunately, most of the other files (e.g. `Autoschedule.h/cpp`, `LoopNest.h/cpp`, etc.) have diverged considerably and it would be a significant amount of work to merge them.

No worries then.

aekul commented:
I think I've now moved all the identical files to `autoschedulers/common`. Are there other things that you think should be done or is this ready for review?

vladl-innopeaktech commented:
Just FYI, I am getting multiple build failures from the latest commit, starting from a fresh environment.

Error log here:
https://gist.github.com/vladl-innopeaktech/6539b811dc2c90bcb965660ea0e8b529

vladl-innopeaktech commented:
Using this patch https://gist.github.com/vladl-innopeaktech/37e5fc044479ca7186911dc140d71479 I am able to whittle down the errors to just some linker failure here:
```
[1/23] Linking CXX executable src/autoschedulers/anderson2021/anderson2021-test_apps_autoscheduler
FAILED: src/autoschedulers/anderson2021/anderson2021-test_apps_autoscheduler 
: && /usr/bin/c++  -O3 -DNDEBUG   src/autoschedulers/anderson2021/CMakeFiles/anderson2021-test_apps_autoscheduler.dir/test.cpp.o  -o src/autoschedulers/anderson2021/anderson2021-test_apps_autoscheduler  -Wl,-rpath,/home/vladl/gpu-autosched/Halide/build/src  src/libHalide.so.15.0.0  -ldl  -pthread && :
/usr/bin/ld: src/autoschedulers/anderson2021/CMakeFiles/anderson2021-test_apps_autoscheduler.dir/test.cpp.o: in function `main':
test.cpp:(.text.startup+0x5e1): undefined reference to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const'
/usr/bin/ld: test.cpp:(.text.startup+0xf49): undefined reference to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const'
/usr/bin/ld: test.cpp:(.text.startup+0x1af9): undefined reference to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const'
/usr/bin/ld: test.cpp:(.text.startup+0x25d1): undefined reference to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const'
/usr/bin/ld: test.cpp:(.text.startup+0x2cb3): undefined reference to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const'
/usr/bin/ld: src/autoschedulers/anderson2021/CMakeFiles/anderson2021-test_apps_autoscheduler.dir/test.cpp.o:test.cpp:(.text.startup+0x2f03): more undefined references to `Halide::Pipeline::auto_schedule(Halide::Target const&, Halide::MachineParams const&) const' follow
collect2: error: ld returned 1 exit status
```

steven-johnson commented:
I think I know what is going on -- I will apply your patch and then work on a fix

steven-johnson commented:
@aekul -- please sync this branch with Halide main, then apply the patch that @vladl-innopeaktech offered in the gist above; I will have an additional patch to apply on top of his (working on it now) but I don't have permission to push directly to your fork.

steven-johnson commented:
Please note: the CMake file uses `add_halide_library(... TARGETS cmake)` in a lot of places; this cannot *possibly* be right for this scheduler, since it requires a GPU feature, but `TARGETS cmake` is defined as `The meta-triple `cmake` is equal to the `arch-bits-os` of the current CMake target` ... i.e., no GPU feature will *ever* be present.

steven-johnson commented:
OK, per my previous comment, after you sync this to main and apply the other patch, please apply https://gist.github.com/steven-johnson/263f0fafd34ebbac052873d6340ab56f on top of that -- it gets everything building an (apparently) working on my system (though some work surely remains to be done).

On top of getting it building and passing tests, I also (mostly) finished the work to remove use of env vars for settings, moving them into the AutoschedulerParams setup that is now used in mainline Halide. I think I got all this right. 

(EDIT: oops, I forgot to update the `.sh` files to set the params explicitly rather than setting the old env vars... I'll do that as a followup, or if you like, you can do so, using the scripts from Adams2019 as a model.)

aekul commented:
Thanks for the patches. Synced with main and applied both of them.

Regarding `TARGETS cmake`, is there a way to specify `cuda` as part of the target? I've mostly tested the autoscheduler using the Makefile (i.e. using `host-cuda` for the target) so I'm less familiar with Halide's use of cmake.

steven-johnson commented:
> Thanks for the patches. Synced with main and applied both of them.

...are you sure? The GitHub page says there is an unresolved conflict in src/autoschedulers/adams2019/CMakeLists.txt

> Regarding `TARGETS cmake`, is there a way to specify `cuda` as part of the target? 

adding @alexreinking as I don't know the answer here. Be aware, though, that adding `cuda` specifically in the build file is problematic in that Cuda isn't available on all systems -- I presume this autoscheduler isn't cuda-specific and would work with e.g. Metal, D3D12Compute as well.

> I've mostly tested the autoscheduler using the Makefile (i.e. using `host-cuda` for the target) so I'm less familiar with Halide's use of cmake.

CMake is the build system of record; Make support is only on a best-effort basis. I highly recommend that you consider working with the CMake build system to reduce the number of build glitches you encounter.

alexreinking commented on src/autoschedulers/adams2019/CMakeLists.txt:
```diff
@@ -2,6 +2,9 @@
 ##
 
+set(COMMON_DIR "${Halide_SOURCE_DIR}/src/autoschedulers/common")
+include_directories("${Halide_SOURCE_DIR}/src/autoschedulers/adams2019")
```
We do not use `include_directories` in this codebase. Please read `README_cmake.md`.

alexreinking replied:
I doubt this is necessary to begin with.

skhiat commented:
Testing the head I got a:
'Error: Constraint violated: head2_filter.extent.1 (73) == 39 (39)'
Here the repro:
https://gist.github.com/skhiat/fe363b516143ee9f47b7b13822e214fe // this code are using D3D12Compute, tested with CUDA same result
For a trivial separated box-filter-3x3

alexreinking commented:
> Be aware, though, that adding `cuda` specifically in the build file is problematic in that Cuda isn't available on all systems -- I presume this autoscheduler isn't cuda-specific and would work with e.g. Metal, D3D12Compute as well.

The argument of interest to `add_halide_library` is `FEATURES cuda cuda_capability_XX` where `XX` is the minimum CUDA capability number you need (e.g. `50`). But as @steven-johnson says, this isn't appropriate for the main build.

The autoscheduler itself presumably does not need CUDA to build. That's the part that belongs in `src/autoscheduler/...`. Tests belong in `tests/autoscheduler/<autoscheduler-name>` and may have additional dependencies.

steven-johnson commented:
> That's the part that belongs in src/autoscheduler/.... Tests belong in tests/autoscheduler/<autoscheduler-name> and may have additional dependencies.

That's beyond the scope of this PR, unfortunately. We should really refactor the other autoscheduler tests in this way first so that there's an existing framework to work with.

alexreinking commented:
> That's beyond the scope of this PR, unfortunately. We should really refactor the other autoscheduler tests in this way first so that there's an existing framework to work with.

Maybe so, but maybe that just means this PR has to wait until that refactoring is done. It will be at least as much work to add an optional cuda dependency here.

skhiat commented:
Fix localy:
```
- add_executable(adams2019_weightsdir_to_weightsfile ${COMMON_DIR}/weightsdir_to_weightsfile.cpp Weights.cpp)
+ add_executable(adams2019_weightsdir_to_weightsfile ${COMMON_DIR}/weightsdir_to_weightsfile.cpp ${COMMON_DIR}/Weights.cpp)
```
// I don't have the right to push on this branch

ryanstout commented:
Any more updates on this? Years in the making here :-) Would love to see this land some day. Thanks!

aekul commented:
Looks like the test reorganization mentioned above has been done. I'll try to take another look at this in the next few days.

ryanstout commented:
@aekul thanks! It would help me out a ton if it landed.

vladl-innopeaktech commented:
@aekul It looks like the autotuning script and the `retrain_cost_model` binary are out-of-sync. I've tried to bridge the gap with the arguments but it looks like `retrain_cost_model` just gets stuck on loading samples forever. Are they very far out of sync? (i.e. should I keep trying to get it working or leave it to you?)

aekul commented:
Haven't had a chance to test the scripts since my last commits but I'll take a look soon.

vladl-innopeaktech commented:
Ok. My team is very interested in getting the autotuning running and I've been playing with this branch for the last couple of months so if there's anything you'd like to delegate (esp wrt the build system) in this PR, I can get time to work on it.

aekul commented:
The one issue I know of with the autotuning scripts is that they use the Makefiles to build Halide and the autoscheduler. So they likely need to be ported to use CMake. Does that sound like the issue you are encountering or is it something else?

aekul commented:
I've updated the autotuning/data generation scripts to use the CMake build and tidied some things up. It should now be possible to build everything with CMake and run the scripts to autotune a generator. There are 3 things to be aware of:
1. `autotune_loop.sh` links `RunGenMain.o` into each benchmark binary separately from the output of the autoscheduler. I didn't see a way to build `RunGenMain.o` with CMake so I built it with the Makefiles and copied to to `path/to/build/tools`
2. If you want to autotune the apps, they need to be compiled. The root `CMakeLists.txt` does not automatically compile the apps. I added `add_subdirectory(apps)` to it and had to comment out `add_app(hannk)` from `apps/CMakeLists.txt`. I haven't checked in these changes because I'm not sure if `add_subdirectory(apps)` was omitted intentionally.
3. There is currently a buildbot error that `included_schedule_file.schedule.h` does not exist. I can't currently check it in because a rule in `.gitignore` is preventing it. So in order to build things, it's currently necessary to add this file locally (e.g. by copying it from the corresponding `adams2019` directory).

@steven-johnson Do you have thoughts/suggestions on any of the above, particularly for (3) and whether I should override that rule and check in the file or do something else?

alexreinking commented:
> 2\. I haven't checked in these changes because I'm not sure if `add_subdirectory(apps)` was omitted intentionally.

It is omitted intentionally.

>  I can't currently check it in because a rule in .gitignore is preventing it.

You can override with `git add -f` (force). That rule exists to prevent accidentally checking in compiler outputs.

NewProggie commented on src/autoschedulers/anderson2021/autotune_loop.sh:
```diff
@@ -0,0 +1,581 @@
+
+RUNGENMAIN="${TOOLS_BUILD_DIR}/RunGenMain.o"
+if [ ! -f $RUNGENMAIN ]; then
+    echo "RunGenMain.o not found. Exiting..."
```
This is problematic as it will fail when building with CMake: CMake generates a `RunGenMain.cpp.o` as opposed to Make which generates a `RunGenMain.o`

NewProggie commented on src/autoschedulers/anderson2021/autotune_loop.sh:
```diff
@@ -0,0 +1,581 @@
+# benchmarked serially.
+BATCH_SIZE=80
+EPOCHS=200
+NUM_GPUS=$(nvidia-smi --query-gpu=name --format=csv,noheader | wc -l)
```
This requires the auto tuning of Nvidia graphic cards specifically, which is probably fine, but I would add an error message in case somebody tries this with any other GPU vendor.

steven-johnson replied:
Yes, definitely, NVidia is not the only GPU vendor out there :-)

NewProggie commented on src/autoschedulers/anderson2021/scripts/utils.sh:
```diff
@@ -0,0 +1,323 @@
+
+function make_dir_path_absolute() {
+    local -r path=$1
+    local -n absolute_path_ref=$2
```
This will fail on recent macOS as they've switched to zsh as their default, which doesn't support '-n'. Consider adding another warning for the user?

NewProggie commented on src/autoschedulers/anderson2021/generate_data.sh:
```diff
@@ -0,0 +1,230 @@
+#
+# Arguments:
+# halide_build_dir - path where Halide is built
+# max_iterations - the number of batches to generate. The cost model is
```
Maybe add a set of valid inputs for each argument to this shell script, e.g. `train_only [0|1] - don't generate...` or something along the line as it makes using this script a little easier

NewProggie commented on src/autoschedulers/anderson2021/generate_data.sh:
```diff
@@ -0,0 +1,230 @@
+    echo "Predict only mode: ON"
+fi
+
+DEFAULT_SAMPLES_DIR_NAME="${SAMPLES_DIR:-autotuned_samples}"
```
This will wail if the user has called this script with `train_only == 1` as the executable for retraining the cost model will silently fail loading empty samples and generate a hard to parse error message.

steven-johnson replied:
hard-to-parse error messages are not good, please fix :-)

NewProggie reviewed (commented):
Thanks for your outstanding work. I've added a few review suggestions that you might consider adding before this gets merged back to the main branch.

steven-johnson commented:
Please get the buildbots green before requesting a review :-)

aekul commented:
@steven-johnson One of the tests for Mullapudi2016 (`mullapudi2016_reorder`) is failing. Looks like it's a performance test but I don't see anything in this PR that would impact that. Has this kind of thing happened before?

stevesuzuki-arm commented on src/autoschedulers/anderson2021/CMakeLists.txt:
```diff
@@ -0,0 +1,159 @@
+    $<TARGET_OBJECTS:anderson2021_weights_obj>
+)
+
+target_include_directories(Halide_Anderson2021 PUBLIC "${Halide_SOURCE_DIR}/src/autoschedulers/anderson2021")
```
I'm very interested in this work and have started to play around, so let me share some findings.
You might be aware already, but, in order to make Halide::Anderson2021 available in installed Halide package, you would need to add it to packaging [here](https://github.com/halide/Halide/blob/147ff481ee5495c21253db58d480e411cfb5b9f1/packaging/CMakeLists.txt#L35). In addition, I needed to modify `PUBLIC` to `PRIVATE` in this line to fix CMake error.

stevesuzuki-arm commented on src/autoschedulers/anderson2021/FunctionDAG.cpp:
```diff
@@ -0,0 +1,1295 @@
+                // Get the bounds estimate
+                map<string, Span> estimates;
+                for (const auto &b : consumer.schedule().estimates()) {
+                    int64_t i_min = *as_const_int(b.min);
```
[#7321](https://github.com/halide/Halide/pull/7321/files#diff-9337dd7a81eff9f8739f231412b22a0bdbfa4adb92a5cc125e5a81ea739ad413) would apply.

soufianekhiat commented:
Hello I tested a simple Histogram with this branch:
```
H::ImageParam	src{ H::UInt(8), 3, "src" };
H::Param<int>	bin{ "bin", 1024 };

H::RDom imgDom(src);
H::Var z("z");
hist(z) = H::cast<UInt32>(0);
hist(H::clamp(H::cast<Int32>(src(imgDom[0], imgDom[1], imgDom[2])), 0, 255)) += H::cast<UInt32>(1);
```

Anderson2021 autoschedule give me:
```
Best cost: 6.38469
auto pipeline = get_pipeline();
Func hist = pipeline.get_func(1);
Var z(hist.get_schedule().dims()[0].var);
Var zi("zi");
RVar src_x(hist.update(0).get_schedule().dims()[0].var);
RVar src_y(hist.update(0).get_schedule().dims()[1].var);
RVar src_z(hist.update(0).get_schedule().dims()[2].var);
Var zi_serial_outer("zi_serial_outer");
hist.update(0)
    .reorder(src_x, src_y, src_z)
    .gpu_single_thread();
hist
    .split(z, z, zi, 32, TailStrategy::ShiftInwards)
    .compute_root()
    .reorder(zi, z)
    .gpu_blocks(z)
    .split(zi, zi_serial_outer, zi, 32, TailStrategy::GuardWithIf)
    .gpu_threads(zi);

produce hist:
  gpu_block z.z<Default_GPU>:
    gpu_thread z.zi.zi in [0, 31]<Default_GPU>:
      hist(...) = ...
  gpu_block __outermost.__outermost.v39 in [0, 0]<Default_GPU>:
    gpu_thread __outermost.v40 in [0, 0]<Default_GPU>:
      for src:
        for src:
          for src:
            hist(...) = ...
```

And Li2018 give me:
```
hist.compute_root()
    .split(z,v3,v4,64,ShiftInwards)
    .reorder(v4,v3)
    .gpu_blocks(v3)
    .gpu_threads(v4)
;
hist.update(0)
    .split(src$y,r18,r19,48,GuardWithIf)
    .split(src$z,r20,r21,40,GuardWithIf)
;
hist_intm = hist.update(0)
    .rfactor({{r18,v11},{r20,v12}})
    .compute_root()
    .split(v11,v11,v21,32,GuardWithIf)
    .fuse(v11,v12,v11)
    .reorder(v21,v11)
    .gpu_blocks(v11)
    .gpu_threads(v21)
;
hist_intm.update()
    .split(v11,v11,v22,32,GuardWithIf)
    .fuse(v11,v12,v11)
    .fuse(src$x,r19,src$x)
    .fuse(src$x,r21,src$x)
    .reorder(v22,v11,src$x)
    .gpu_blocks(v11)
    .atomic()
    .gpu_blocks(src$x)
    .gpu_threads(v22)
hist.update(0)
    .split(z,v31,v32,64,GuardWithIf)
    .reorder(r18,r20,v32,v31)
    .gpu_blocks(v31)
    .gpu_threads(v32)
;
src_im.compute_root()
    .split(_1,v35,v36,64,ShiftInwards)
    .fuse(_0,v35,_0)
    .fuse(_0,_2,_0)
    .reorder(v36,_0)
    .reorder_storage(_1,_0,_2)
    .gpu_blocks(_0)
    .gpu_threads(v36)
;

produce src_im:
  gpu_block _0._0._0<Default_GPU>:
    gpu_thread _1.v36 in [0, 63]<Default_GPU>:
      src_im(...) = ...
consume src_im:
  produce hist_intm:
    gpu_block v11.v11.v11<Default_GPU>:
      gpu_thread v11.v21 in [0, 31]<Default_GPU>:
        for z:
          hist_intm(...) = ...
    gpu_block src.src.src<Default_GPU>:
      gpu_block v11.v11.v11<Default_GPU>:
        gpu_thread v11.v22 in [0, 31]<Default_GPU>:
          hist_intm(...) = ...
  consume hist_intm:
    produce hist:
      gpu_block z.v3<Default_GPU>:
        gpu_thread z.v4 in [0, 63]<Default_GPU>:
          hist(...) = ...
      gpu_block z.v31<Default_GPU>:
        gpu_thread z.v32 in [0, 63]<Default_GPU>:
          for src.r20:
            for src.r18:
              hist(...) = ...
```

What surprise me is Anderson2021 generate 1x1 Dispatch in a single thread and Li2018 generate proper parallel compute with multiple thread.

aekul commented:
Several things to consider:
- Do those parameters correspond to your GPU's specifications?
- Did you autotune the app or just run it once? Autotuning is generally the best way to find a schedule for a single app
- It's always possible that the model predicts the cost of a schedule inaccurately and gives suboptimal results

soufianekhiat commented:
> Do those parameters correspond to your GPU's specifications?
Yes
> Did you autotune the app or just run it once? Autotuning is generally the best way to find a schedule for a single app
I autotuned it with 100 passes

aekul commented:
Increasing the number of passes does not autotune it (it does not retrain the cost model). You should try using `generate_data.sh` if you haven't already.

aekul commented:
@steven-johnson The buildbots are green now

steven-johnson commented on src/autoschedulers/anderson2021/AutoSchedule.cpp:
```diff
@@ -0,0 +1,821 @@
+    IntrusivePtr<State> optimal_schedule(int beam_size);
+};
+
+#ifdef HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API
```
FYI: HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API is now removed in Halide `main`, so this should just be removed here and elsewhere. (We can do this as a followup PR if you like.)

steven-johnson commented on src/autoschedulers/anderson2021/AutoSchedule.cpp:
```diff
@@ -0,0 +1,821 @@
+  That said, here are the (legacy) env vars you can still use when HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API
+  is defined:
+
+  HL_BEAM_SIZE
```
FYI: since HALIDE_ALLOW_LEGACY_AUTOSCHEDULER_API is now removed in Halide `main`, most of these env vars are probably no longer actually in use (settings would be controlled by GeneratorParams instead), so this should probably get pruned (subsequent PR is fine)

steven-johnson commented on src/autoschedulers/anderson2021/AutoSchedule.h:
```diff
@@ -0,0 +1,27 @@
+
+typedef PerfectHashMap<FunctionDAG::Node::Stage, ScheduleFeatures> StageMapOfScheduleFeatures;
+
+void find_and_apply_schedule(FunctionDAG &dag, const std::vector<Function> &outputs, const Anderson2021Params &params, const Target &target, CostModel *cost_model, int beam_size, StageMapOfScheduleFeatures *schedule_features);
```
Style nit: this is an awfully long line

steven-johnson commented on src/autoschedulers/anderson2021/DefaultCostModel.cpp:
```diff
@@ -0,0 +1,401 @@
+                  "Incorrect size for pipeline features");
+    int num_stages = 0;
+    for (const auto &n : dag.nodes) {
+        if (!n.is_input) num_stages += (int)n.stages.size();
```
style nit, here and elsewhere: generally, Halide always encloses if clauses in braces, even for a single line:

```
if (...) {
   ...
}
```

steven-johnson commented on src/autoschedulers/anderson2021/DefaultCostModel.cpp:
```diff
@@ -0,0 +1,401 @@
+        *(cost_ptrs(i)) = dst(i);
+        if (std::isnan(dst(i))) {
+            any_nans = true;
+            aslog(0) << "Prediction " << i << " is NaN. True runtime is " << true_runtimes(i) << "\n";
```
Generally, we prefer the autoschedulers to be 'quiet' by default -- do we really want/need these to be `aslog(0)` (vs `aslog(1)` etc)?

steven-johnson commented on src/autoschedulers/anderson2021/DefaultCostModel.cpp:
```diff
@@ -0,0 +1,401 @@
+            any_nans = true;
+            aslog(0) << "Prediction " << i << " is NaN. True runtime is " << true_runtimes(i) << "\n";
+            aslog(0) << "Checking pipeline features for NaNs...\n";
+            pipeline_feat_queue.for_each_value([&](float f) { if (std::isnan(f)) abort(); });
```
Prefer to use `internal_assert` or `user_assert` rather than a naked `abort()`

steven-johnson commented on src/autoschedulers/anderson2021/DefaultCostModel.h:
```diff
@@ -0,0 +1,77 @@
+          stats{stats} {
+        load_weights();
+    }
+    ~DefaultCostModel() override = default;
```
...why do we need to override it if the implementation is default?

steven-johnson commented on src/autoschedulers/anderson2021/FunctionDAG.cpp:
```diff
@@ -0,0 +1,1298 @@
+    unsigned char *mem = (unsigned char *)malloc(bytes_to_allocate);
+
+    blocks.push_back(mem);
+    static_assert((sizeof(BoundContents) & 7) == 0, "BoundContents header is not aligned");
```
Are you trying to ensure that this is aligned to size of a pointer? If so, do that explicitly: `(sizeof(BoundContents) % sizeof(void*)) == 0`

EDIT: I see that the same pattern is in the adams2019 version too, but it is also naughty, we'll fix it later

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.cpp:
```diff
@@ -0,0 +1,4001 @@
+    return {child, parent};
+}
+
+int64_t LoopNest::points_accessed_per_thread(const Anderson2021Params &params, const Target &target, const GPULoopInfo &gpu_loop_info, const std::vector<const FunctionDAG::Edge *> &edge_chain, const LoadJacobian &jac, const LoopNest *parent, const LoopNest *grandparent, int64_t n, const ScheduleFeatures &feat, const LoadJacobian &serial_jac, bool producer_has_been_scheduled, int producer_innermos ...
```
Line too long

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.cpp:
```diff
@@ -0,0 +1,4001 @@
+                                const LoopNest *parent,
+                                const LoopNest *grandparent,
+                                const LoopNest &root,
+                                int64_t *working_set,
```
It's unusual to have out-params stuck in the middle of the argument list (rather than at the end).

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.cpp:
```diff
@@ -0,0 +1,4001 @@
+                ++stats.num_memoization_misses;
+            }
+
+            c->compute_features(dag, params, target, sites, subinstances, parallelism, this, parent, root, &working_set_here, &working_set_here_local_constant, &working_set_here_local_dynamic, features, gpu_loop_info, use_memoized_features, total_shared_mem_alloc_sizes, stats, verbose);
```
Line too long

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.cpp:
```diff
@@ -0,0 +1,4001 @@
+    ScheduleFeatures &feat = features->get_or_create(stage);
+
+    if (innermost) {
+    } else {
```
Why is this clause empty? Either remove it entirely or add a comment explaining why it is empty

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.cpp:
```diff
@@ -0,0 +1,4001 @@
+
+    if (innermost && !is_scalar()) {
+        compute_warp_features(feat, gpu_loop_info);
+
```
remove blank line

steven-johnson commented on src/autoschedulers/anderson2021/LoopNest.h:
```diff
@@ -0,0 +1,601 @@
+template<typename T>
+using StageMap = PerfectHashMap<FunctionDAG::Node::Stage, T>;
+
+enum GPU_parallelism { block,
```
Style nit: enums in Halide generally begin with Uppercase. 

Also, using `enum class` tends to be preferred for most new enums at this point (though this definitely isn't required, sometimes terseness or other usage factors makes `enum class` undesirable.

steven-johnson commented on src/autoschedulers/anderson2021/State.cpp:
```diff
@@ -0,0 +1,1301 @@
+    StageMap<StageMap<bool>> descendants;
+    root->get_stages_computed_in_each_compute_root_loop(descendants);
+
+    aslog(0) << "BEGIN compute locations\n";
```
Does this need to be `aslog(0)`?

steven-johnson commented on src/autoschedulers/anderson2021/ThreadInfo.h:
```diff
@@ -0,0 +1,269 @@
+namespace Internal {
+namespace Autoscheduler {
+
+#define MAX_THREADS_PER_BLOCK 1024
```
prefer `constexpr int` for consts like this

steven-johnson commented on src/autoschedulers/anderson2021/Tiling.h:
```diff
@@ -0,0 +1,57 @@
+#include <cstdint>
+#include <vector>
+
+using std::vector;
```
Never, ever, ever put a `using` into a .h file in the global namespace.

steven-johnson reviewed (approved):
This looks pretty great, thanks for doing all this work to get this landed! Just a handful of mostly style nits and such, with a few relatively minor things that need addressing (and a few things that would be nice for someone to work on as a followup).

Yongqi-Zhuo commented on src/autoschedulers/anderson2021/DefaultCostModel.cpp:
```diff
@@ -0,0 +1,401 @@
+    }
+
+    cursor = 0;
+    cost_per_stage_ptrs.clear();
```
I might have been wrong, but it seems that the buildbots are not running the GPU autoscheduler related tests?
For example, I pulled this branch and tried running `anderson2021_test_apps_autoscheduler` (i.e., `test/autoschedulers/anderson2021/test.cpp`) but it instantly fails, complaining that access to `cost_per_stage_ptrs` is out of bound. I commented this line (`cost_per_stage_ptrs.clear();`) and it seems to work again.

steven-johnson replied:
> I might have been wrong, but it seems that the buildbots are not running the GPU autoscheduler related tests?

That's correct and as-intended, since this branch hasn't landed yet. The buildbots usually aren't configured to test major new features before they have landed. This PR is close, but has a number of nits that need fixing first.

Yongqi-Zhuo commented on src/autoschedulers/anderson2021/State.cpp:
```diff
@@ -0,0 +1,1301 @@
+            continue;
+        }
+
+        // Make the vectorized dimension of the inner loop 32 (or as
+        // close as possible)
+        int64_t inner_extent = std::min(c->size[vectorized_loop_index], (int64_t)32);
+
+        if (c->stage->index == 0) {
+            vector<int64_t> tiling(c->node->dimensions, 1);
+
+            // Split into parallelized and serial
+            c = c->parallelize_in_tiles(tiling, loop_nest, params, target, true, false);
+
+            if (vectorized_loop_index >= 0) {
+                tiling[vectorized_loop_index] = inner_extent;
+            }
```
It seems that it is not checked for whether `vectorized_loop_index == -1` here in line 222 (and it seems that `-1` is a legal value here because of the subsequent check in line 230), and leads to a crash when running `test/autoschedulers/anderson2021/test.cpp`. I changed
```C++
int64_t inner_extent = std::min(c->size[vectorized_loop_index], (int64_t)32);
```
to
```C++
int64_t inner_extent = vectorized_loop_index != -1 ? std::min(c->size[vectorized_loop_index], (int64_t)32) : (int64_t)32;
```
to fix this but I am not sure if this is right.

aekul commented on src/autoschedulers/anderson2021/FunctionDAG.cpp:
```diff
@@ -0,0 +1,1298 @@
+        } else if (op->call_type == Call::Image) {
+            visit_memory_access(op->name, op->type, op->args, PipelineFeatures::AccessType::LoadImage);
+            op_bucket(PipelineFeatures::OpType::ImageCall, op->type)++;
+        }  // TODO: separate out different math calls a little better (sqrt vs sin vs lerp)
```
Adams2019 has the same TODO. Might be worth tracking

aekul commented:
@steven-johnson Thanks for the review! Made all the requested changes.

Thanks to all the others who left comments/suggestions too.

abadams commented on apps/images/low_res_in.png:
This appears to be a duplicate of apps/images/rgb_small.png

abadams commented on test/autoschedulers/adams2019/demo_generator.cpp:
```diff
@@ -6,10 +6,10 @@ using namespace Halide;
 
 class ConvRelu : public Halide::Generator<ConvRelu> {
 public:
-    Input<Buffer<float, 4>> input{"input"};
```
These changes look unintentional (same for the file below)

aekul replied:
Fixed here: https://github.com/halide/Halide/pull/7475
